---
title: "scRNA-seq Summary Report"
output:
  pdf_document:
    toc: true
    number_sections: true
    fig_caption: true
    latex_engine: xelatex
    includes:
      in_header: common-pdf.tex
params:
  config.args: "./config.yaml"
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(
  warning=FALSE,error=FALSE, message=FALSE, echo=FALSE
)

source("functions_for_report.R") # This is the file that holds the R functions used in the RMD file
```

```{r libraries}
if (!suppressMessages(require(Seurat))) stop("Seurat is required")
if (!suppressMessages(require(yaml))) stop("yaml is required")
if (!suppressMessages(require(cluster))) stop("cluster is required") # For silhouette score calculation
if (!suppressMessages(require(future))) stop("future is required")
if (!suppressMessages(require(future.apply))) stop("future.apply is required")
if (!suppressMessages(require(cowplot))) stop("cowplot is required")
if (!suppressMessages(require(parallelly))) stop("parallelly is required")
if (!suppressMessages(require(ggplot2))) stop("ggplot2 is required")
if (!suppressMessages(require(DoubletFinder))) stop("DoubletFinder is required")


# the presto lib isn't required, but FindAllMarkers runs MUCH faster with it. 
# For example, for 80k cells and 10 clusters, without presto FindAllMarkers 
# took more than 5 hours, but with presto loaded it took less than 
# 5 minutes!!
#
# The latest ragg, which is a dependency for devtools, which is itself required
# for presto install, doesn't work with R/4.4.0 (as of 13-Nov-2025), so install 
# an older version of ragg, then install devtools, then install presto
# 
# install.packages(
#   "https://cran.r-project.org/src/contrib/Archive/ragg/ragg_1.4.0.tar.gz",
#   repos = NULL,
#   type = "source"
# )
# install.packages(devtools)
# devtools::install_github("immunogenomics/presto")
if (!suppressMessages(require(presto))) stop("presto is required")


```

```{r prepare_inline_variables, eval = TRUE}
# get the datetime and set it to variable
todays.date <- Sys.Date()
formatted.date <- format(todays.date, format="%B %d %Y")

config.args <- read_yaml(params$config.args)

```

```{r user_vals, eval=TRUE}
cfg <- config.args
cfg.analysis <- cfg$analysis

p1  <- cfg.analysis$part1
p2 <- cfg.analysis$part2

r.obj.loc <- cfg$data$`rds-file-path`
organism <- cfg$project$organism

################################################################################
# QC cutoffs
################################################################################

mito.ceiling = p2$mito_ceiling
RNA.count.floor = p2$RNA_count_floor
RNA.count.ceiling = p2$RNA_count_ceiling
feature.count.floor = p2$feature_count_floor
feature.count.ceiling = p2$feature_count_ceiling
rbc.ceiling = p2$rbc_ceiling
ribo.floor = p2$ribo_floor

node.type <- cfg.analysis$node_type

################################################################################
# I/O paths
################################################################################
part1.suffix <- p1$part1_suffix
part2.suffix <- p2$part2_suffix

integrated.seurat.obj.path <- suffix_path(r.obj.loc, p2$integrated_seurat_obj_file, part2.suffix)
QC.tables.path         <- suffix_path(r.obj.loc, p1$part1_report_tables_filename, part1.suffix)
ggplot.directory       <- file.path(r.obj.loc, p1$ggplot_dir)
marker.gene.dir <- suffix_path(r.obj.loc, p2$marker_gene_dir, part2.suffix)

  if (!file.exists(ggplot.directory)) {
 dir.create(ggplot.directory, showWarnings = FALSE, recursive = TRUE)
}

if (!file.exists(marker.gene.dir)) {
 dir.create(marker.gene.dir, showWarnings = FALSE, recursive = TRUE)
}
```


## Report and variable details

This report was generated on `r formatted.date`

```{r get_num_cores_available}
if (node.type == "compute"){
  # If using a compute node, then put this in the submission script:
  # export MC_CORES=${SLURM_NTASKS}
  num.cores.available <- as.numeric(Sys.getenv("SLURM_NTASKS"))
} else {
  library(parallel)
  num.cores.available <- detectCores() 
}
```


```{r load_data}
seurat5.obj <- readRDS(file=integrated.seurat.obj.path)
```

```{r group_assignment}
# example
# obj$group <- NA_character_
# obj$group[grepl("OA$", obj$sample_id)] <- "OA"
# obj$group[grepl("VE$", obj$sample_id)] <- "VE"


seurat5.obj$sex <- "M"
seurat5.obj$sex[grepl("_F_", seurat5.obj$sample_id)] <- "F"
```


```{r find_markers_sct}
DefaultAssay(seurat5.obj) <- "SCT"
Idents(seurat5.obj) <- "sex"
# perform differential expression

# seurat5.obj <- PrepSCTFindMarkers(seurat5.obj)

# Runs MUCH faster with presto package installed
sex.markers <- FindAllMarkers(seurat5.obj, 
                                      only.pos = TRUE, 
                                      verbose = TRUE)

saveRDS(sex.markers, file=file.path(marker.gene.dir, "sex_markers_sct.RDS"))
 
male.markers.path <- file.path(marker.gene.dir, "male_markers.csv")
male.markers <- sex.markers[sex.markers$cluster == "M",] 
write.csv(x = male.markers, file = male.markers.path, quote = FALSE)

female.markers.path <- file.path(marker.gene.dir, "female_markers.csv")
female.markers <- sex.markers[sex.markers$cluster == "F",] 
write.csv(x = female.markers, file = female.markers.path, quote = FALSE)

write.csv(sex.markers, 
          file.path(marker.gene.dir,"sex_markers.csv"))
```

```{r sex_dimplot}
DimPlot(seurat5.obj, 
        reduction = "umap.postint") + ggtitle("Cells by sex")

```

## Session Information
```{r sessioinfo}
sessionInfo()
```

