---
output:
  html_document:
    pandoc_args: './common.yaml'
params:
    config.args: "./config.yaml"
---
```{r import_libraries, warning=FALSE,error=FALSE,message=FALSE}
# must have Seurat
suppressMessages(require(Seurat))
suppressMessages(require(Matrix))
suppressMessages(require(kableExtra))
suppressMessages(require(ggplot2))
suppressMessages(require(yaml))

```


```{r user_vars, echo = FALSE}

# get the datetime and set it to variable for inline printing
todays.date <- Sys.Date()
formatted.date <- format(todays.date, format="%B %d %Y")

# Import the config file
config.args <- read_yaml(params$config.args)

run_metrics_tables=TRUE
metrics.table.string <- config.args$analysis$part1$generate_metrics_tables

if (metrics.table.string == "FALSE"){
  run_metrics_tables=FALSE
}


# set project variables
analyst.name <- config.args$project$`analyst-name`
formatted.samples <- config.args$project$samples
experiment.name = config.args$project$`project-name` 
project.description <- config.args$project$`project-decription`
dataset.path <- config.args$data$parent_directory_path
r.obj.loc <- config.args$data$`rds-file-path` # Output folder

# When trying different parameters in the analysis you can create
# alternate output files with this 'suffix' argument in the config
part1.suffix <- config.args$analysis$part1$part1_suffix

part1.rds.save.file <- config.args$analysis$part1$part1_rds_save_filename
part1.rds.save.file <- sub(".RDS", paste0(part1.suffix, ".RDS"), part1.rds.save.file )

part1.rds.save.path <- file.path(r.obj.loc, part1.rds.save.file)

part1.file.for.report <- config.args$analysis$part1$part1_report_tables_filename
part1.file.for.report <-sub(".RData", paste0(part1.suffix, ".RData"), part1.file.for.report)

part1.file.report.path <-file.path(r.obj.loc, part1.file.for.report)

ggplot.directory.name <- config.args$analysis$part1$ggplot_dir
ggplot.directory <- file.path(r.obj.loc, ggplot.directory.name)

# create ggplot folder if it doesn't exist already
if (!file.exists(ggplot.directory)) {
 dir.create(ggplot.directory, showWarnings = FALSE, recursive = TRUE)
}


# create rds folder if it doesn't exist already
if (!file.exists(r.obj.loc)) {

 dir.create(r.obj.loc, showWarnings = FALSE, recursive = TRUE)

}

# setup the data location variables
import.h5 <- config.args$data$`import-h5` # set to false if you want to import the matrix files


organism = config.args$project$organism

# setup the sample names for printing in the inline report
formatted.samples <- gsub("\"", "", formatted.samples)
formatted.samples <- gsub("\ ", "  \n ", formatted.samples)

# These two replace the first dash in the file names, which comes from the 
# formatting of the config file, with '*' for the rmd report
formatted.samples <- gsub("\ -", " * ", formatted.samples)
formatted.samples <- gsub("^-", " * ", formatted.samples)

formatted.samples <- paste0(formatted.samples, " <br /> ")

# generate list of ids for use in importing data
ids <- config.args$project$samples # sample names, need to match the folders containing the h5 files
seurat.sample.names <- gsub("-", "_", ids)

# Use sub instead of gsub in the first call. gsub replaces all, while sub 
# only replaces the first instance. The first dash is just there as formatting
# in the config file. This allows for names with dashes in them.
ids <- sub("^-", "", ids)
ids <- gsub("\ -", "\ ", ids)
ids <- gsub("\"", "", ids)
ids <- strsplit(ids, "[[:space:]]")
ids <- unlist(ids)

if (organism == "human"){
  mt.marker = "^MT-"
  hb.marker = "^HB[^(P)]"
  ribo.marker = "^RP[SL]"
  xist.gene = "XIST"
  genes.file = "/projects/b1012/xvault/REFERENCES/10xGenomics/RNA-seq/hg38/refdata-gex-GRCh38-2020-A/genes/genes.gtf"
} else if (organism == "mouse") {
  mt.marker = "^mt-"
  hb.marker = "^Hb[^(p)]"
  ribo.marker = "^Rp[sl]" 
  genes.file = "/projects/b1012/xvault/REFERENCES/10xGenomics/RNA-seq/mm10/refdata-gex-mm10-2020-A/genes/genes.gtf"
  xist.gene = "Xist"
} else {
  mt.marker = "NONE"
  hb.marker = "NONE"
  ribo.marker = "NONE"
  xist.gene = "NONE"
}
```

```{r define_save_plots, echo = FALSE}

save_png_plot <- function(plot, filename){
  save.path <- file.path(ggplot.directory, paste0("part1", part1.suffix, "_", filename))
  ggsave(save.path, plot)
}

```

## Report and variable details

* This report was generated on `r formatted.date`

* starting with raw data at `r dataset.path`
* saving rds object to `r part1.rds.save.path`
* saving file for report at `r part1.file.report.path`

* organism: `r organism`
* Importing h5?: `r import.h5`

# Project Description:

`r project.description`

<br/>
<br/>

## Sample Names

`r formatted.samples `

<br/>
<br/>

# Part 1: Loading data from CellRanger into R

Part 1 focuses on getting data into R and setting up our initial object.

<br/>
<br/>

`r if (run_metrics_tables) {"## Read in the cellranger sample metrics csv files"}`

<br/>
<br/>

```{r read_metrics, warning=FALSE,error=FALSE, message=FALSE, eval=run_metrics_tables}
d10x.metrics <- lapply(ids, function(sample.name){
  print(sample.name)
  metrics.path.cleaned <- file.path(dataset.path,sample.name,"metrics_summary.csv")
  metrics.path.raw <- file.path(dataset.path, sample.name, "outs/metrics_summary.csv") 
  tryCatch( 
    {  
      if(file.exists(metrics.path.cleaned)){     
        metrics <- read.csv(metrics.path.cleaned, colClasses = "character")  
        } else if (file.exists(metrics.path.raw)) { 
          metrics <- read.csv(metrics.path.raw, colClasses = "character")  
        } 
      }, 
    error=function(cond) { 
      message(paste0("Error loading the sample ", sample.name)) 
      message("Here's the original error message:")  
      }
    )
})
experiment.metrics <- do.call("rbind", d10x.metrics)
rownames(experiment.metrics) <- ids


# I want the formatting in the charts below to include percent symbols and commas separating thousands, but I also
# want numeric values in order to scale the colors. So, I am setting up both character and numeric copies of the
# metrics, one for printing and one for scaling.
sequencing_metrics <- data.frame(experiment.metrics[,c(4:16,1,17,2,3,18,19)])
sequencing.metrics.numeric <- as.data.frame(lapply(sequencing_metrics, gsub, pattern = ",", replacement = ""))
sequencing.metrics.numeric <- as.data.frame(lapply(sequencing.metrics.numeric, gsub, pattern = "%", replacement = ""))

if (length(ids) > 1){ 
  sequencing.metrics.numeric <- as.data.frame(apply(sequencing.metrics.numeric, 2, as.numeric))
} else {
  sequencing.metrics.numeric <- as.data.frame(lapply(sequencing.metrics.numeric, as.numeric))
}
# sequencing.metrics.numeric
# 
# sapply(sequencing_metrics, class)
# chars <- sapply(sequencing_metrics, is.character)
# sequencing_metrics[ , chars] <- as.data.frame(apply(sequencing_metrics[ , chars], 2, as.numeric))
# sapply(sequencing.metrics.numeric, class)
# sequencing_metrics

row.names(sequencing_metrics) <- gsub("\\."," ", rownames(sequencing_metrics))

sequencing.characteristics.numeric <- sequencing.metrics.numeric[,1:6]
sequencing.characteristics <- sequencing_metrics[,1:6]

mapping.characteristics.numeric <- sequencing.metrics.numeric[,7:13]
mapping.characteristics <- sequencing_metrics[,7:13]

cell.characteristics.numeric <- sequencing.metrics.numeric[,14:19]
cell.characteristics <- sequencing_metrics[,14:19]
```

<br/>
<br/>



<br/>

`r if (run_metrics_tables) {"## Sequencing Characteristics"}`
```{r table_seq_characteristics, warning=FALSE,error=FALSE, message=FALSE, eval=run_metrics_tables}

# scale the columns with percentages from 0:100
# direction = -1 will make it so that the text and background don't blend in, unless the percentages are close to 50%
# I set sequencing saturation text to white because those numbers tend to be closer to 50%
sequencing.char.table <- sequencing.characteristics %>%
  kbl() %>%
  kable_paper(full_width = T) %>%
  column_spec(2) %>%
  column_spec(3, 
              color = spec_color(sequencing.characteristics.numeric$Valid.Barcodes, option = "viridis", direction = -1, scale_from = c(0, 100)), 
              background = spec_color(sequencing.characteristics.numeric$Valid.Barcodes, option="viridis", scale_from = c(0, 100))) %>%
  column_spec(4, color = "white", background = spec_color(sequencing.characteristics.numeric$Sequencing.Saturation, option="viridis", scale_from = c(0, 100))) %>%
  column_spec(5, color = spec_color(sequencing.characteristics.numeric$Q30.Bases.in.Barcode, option = "viridis", scale_from = c(0, 100), direction = -1), 
              background = spec_color(sequencing.characteristics.numeric$Q30.Bases.in.Barcode, option="viridis", scale_from = c(0, 100))) %>%
  column_spec(6, color = spec_color(sequencing.characteristics.numeric$Q30.Bases.in.RNA.Read, option="viridis", scale_from = c(0, 100), direction = -1), 
              background = spec_color(sequencing.characteristics.numeric$Q30.Bases.in.RNA.Read, option="viridis", scale_from = c(0, 100))) %>%
  column_spec(7, color = spec_color(sequencing.characteristics.numeric$Q30.Bases.in.UMI, option="viridis", scale_from = c(0, 100), direction = -1), 
              background = spec_color(sequencing.characteristics.numeric$Q30.Bases.in.UMI, option="viridis", scale_from = c(0, 100)))

sequencing.char.table
# sequencing_metrics %>%
#   kable(caption = 'Cell Ranger Results') %>%
#   pack_rows("Sequencing Characteristics", 1, 6, label_row_css = "background-color: #666; color: #fff;") %>%
#   pack_rows("Mapping Characteristics", 7, 13, label_row_css = "background-color: #666; color: #fff;") %>%
#   pack_rows("Cell Characteristics", 14, 19, label_row_css = "background-color: #666; color: #fff;") %>%
#   column_spec(1, color = spec_color(sequencing_metrics$Number.of.Reads)) %>%
#   kable_styling("striped")
```

<br/>

`r if (run_metrics_tables) {"## Mapping Characteristics"}`
```{r table_map_characteristics, warning=FALSE,error=FALSE, message=FALSE, eval=run_metrics_tables}
# scale the columns with percentages from 0:100
# direction = -1 will make it so that the text and background don't blend in, unless the percentages are close to 50%
mapping.char.table.1 <- mapping.characteristics[, c(1:4)] %>%
  kbl() %>%
  kable_paper(full_width = T) %>%
  column_spec(2, color = spec_color(mapping.characteristics.numeric$Reads.Mapped.to.Genome, option="viridis", direction = -1, scale_from = c(0, 100)), 
              background = spec_color(mapping.characteristics.numeric$Reads.Mapped.to.Genome, option="viridis", scale_from = c(0, 100))) %>%
  column_spec(3,
              color = spec_color(mapping.characteristics.numeric$Reads.Mapped.Confidently.to.Genome, option = "viridis", direction = -1, scale_from = c(0, 100)),
              background = spec_color(mapping.characteristics.numeric$Reads.Mapped.Confidently.to.Genome, option="viridis", scale_from = c(0, 100))) %>%
  column_spec(4,
              color = spec_color(mapping.characteristics.numeric$Reads.Mapped.Confidently.to.Intergenic.Regions, option = "viridis", direction = -1, scale_from = c(0, 100)),
              background = spec_color(mapping.characteristics.numeric$Reads.Mapped.Confidently.to.Intergenic.Regions, option="viridis", scale_from = c(0, 100))) %>%
  column_spec(5, color = spec_color(mapping.characteristics.numeric$Reads.Mapped.Confidently.to.Intronic.Regions, option = "viridis", scale_from = c(0, 100), direction = -1),
              background = spec_color(mapping.characteristics.numeric$Reads.Mapped.Confidently.to.Intronic.Regions, option="viridis", scale_from = c(0, 100)))

mapping.char.table.1
```

 <br/>
 
```{r table_map_characteristics_2, warning=FALSE,error=FALSE, message=FALSE, eval=run_metrics_tables}

# scale the columns with percentages from 0:100
# direction = -1 will make it so that the text and background don't blend in, unless the percentages are close to 50%
mapping.char.table.2 <- mapping.characteristics[, c(5:7)] %>%
  kbl() %>%
  kable_paper(full_width = T) %>%
  column_spec(2, color = spec_color(mapping.characteristics.numeric$Reads.Mapped.Confidently.to.Exonic.Regions, option="viridis", scale_from = c(0, 100), direction = -1),
              background = spec_color(mapping.characteristics.numeric$Reads.Mapped.Confidently.to.Exonic.Regions, option="viridis", scale_from = c(0, 100))) %>%
  column_spec(3, color = spec_color(mapping.characteristics.numeric$Reads.Mapped.Confidently.to.Transcriptome, option="viridis", scale_from = c(0, 100), direction = -1),
              background = spec_color(mapping.characteristics.numeric$Reads.Mapped.Confidently.to.Transcriptome, option="viridis", scale_from = c(0, 100))) %>%
  column_spec(4, color = spec_color(mapping.characteristics.numeric$Reads.Mapped.Antisense.to.Gene, option="viridis", scale_from = c(0, 100), direction = -1),
              background = spec_color(mapping.characteristics.numeric$Reads.Mapped.Antisense.to.Gene, option="viridis", scale_from = c(0, 100)))

mapping.char.table.2
```
<br/>

`r if (run_metrics_tables) {"## Cell Characteristics"}`
```{r table_cell_characteristics, warning=FALSE,error=FALSE, message=FALSE, eval=run_metrics_tables}

# scale the columns with percentages from 0:100
# direction = -1 will make it so that the text and background don't blend in, unless the percentages are close to 50%
# I set sequencing saturation text to white because those numbers tend to be closer to 50%
cell.char.table <- cell.characteristics %>%
  kbl() %>%
  kable_paper(full_width = T) %>%
  column_spec(2) %>%
  column_spec(3,
              color = spec_color(cell.characteristics.numeric$Fraction.Reads.in.Cells, option = "viridis", direction = -1, scale_from = c(0, 100)),
              background = spec_color(cell.characteristics.numeric$Fraction.Reads.in.Cells, option="viridis", scale_from = c(0, 100))) %>%
  column_spec(4) %>%
  column_spec(5) %>%
  column_spec(6) %>%
  column_spec(7)

cell.char.table
```

## Load the Cell Ranger Matrix Data and create the base Seurat object.
Cell Ranger provides a function `cellranger aggr` that will combine multiple samples into a single matrix file. However, when processing data in R this is unnecessary and we can quickly aggregate them in R.

Seurat provides a function `Read10X` and `Read10X_h5` to read in 10X data folder. First we read in data from each individual sample folder. 

Later, we initialize the Seurat object (`CreateSeuratObject`) with the raw (non-normalized data). Keep all cells with at least 200 detected genes. Also extracting sample names, calculating and adding in the metadata mitochondrial percentage of each cell. Adding in the metadata batchid and cell cycle. Finally, saving the raw Seurat object.

## Load the Cell Ranger Matrix Data (hdf5 file) and create the base Seurat object.

Objects created for:

<br/>

```{r load_tenx_data, warning=FALSE,error=FALSE, message=FALSE}

d10x.data <- lapply(ids, function(sample.name){
  print(paste0("Importing data for ", sample.name)) 
  
  if (dir.exists(file.path(dataset.path, sample.name, "outs"))){ 
    tenx.h5.path <- file.path(dataset.path,sample.name,"outs/raw_feature_bc_matrix.h5")
    tenx.matrix.path <- file.path(dataset.path,sample.name,"outs/raw_feature_bc_matrix")
  } else {
    tenx.h5.path <- file.path(dataset.path,sample.name,"raw_feature_bc_matrix.h5")
    tenx.matrix.path <- file.path(dataset.path,sample.name,"raw_feature_bc_matrix")
  }

  
  if (import.h5){
    d10x <- Read10X_h5(tenx.h5.path)
  } else {
    d10x <- Read10X(tenx.matrix.path)
  }
  
  # The 'ids' are meant to mimic the directory names, but having a hyphen in the 
  # sample names is a problem. I fixed it above by creating the 'seurat.sample.names'
  # list, but I have to do it again here for the column names.
  sample.name.fixed <- gsub("-", "_", sample.name)
  
  colnames(d10x) <- paste(
    sapply(strsplit(colnames(d10x),split="-"),'[[',1L),sample.name.fixed ,sep="-")
  d10x
})
names(d10x.data) <- seurat.sample.names

# str(d10x.data)
```


## GEX Barcode Rank Plot

The GEX Barcode Rank Plot for the gene expression data enables one to assess library quality. Ideally there is a steep drop-off separating high UMI count cells from low UMI count background noise.

<br/>

```{r fig_barcode_umi, warning=FALSE,error=FALSE, message=FALSE, fig.width=10}
# cr_filtered_cells <- as.numeric(gsub(",","",as.character(unlist(sequencing_metrics["Estimated.Number.of.Cells",]))))
cr_filtered_cells <- unlist(sequencing.metrics.numeric[,"Estimated.Number.of.Cells"])

barcode.umi.fig.list <- list()

plot_cellranger_cells <- function(ind){
  xbreaks = c(1,1e1,1e2,1e3,1e4,1e5,1e6)
  xlabels = c("1","10","100","1000","10k","100K","1M")
  ybreaks = c(1,2,5,10,20,50,100,200,500,1000,2000,5000,10000,20000,50000,100000,200000,500000,1000000)
  ylabels = c("1","2","5","10","2","5","100","2","5","1000","2","5","10k","2","5","100K","2","5","1M")

  pl1 <- data.frame(index=seq.int(1,ncol(d10x.data[[ind]])), nCount_RNA = sort(Matrix:::colSums(d10x.data[[ind]])+1,decreasing=T), nFeature_RNA = sort(Matrix:::colSums(d10x.data[[ind]]>0)+1,decreasing=T)) %>% ggplot() + 
    scale_color_manual(values=c("grey50","red2","blue4"), labels=c("UMI_Background", "Features", "UMI_Cells"), name=NULL) +
    ggtitle(paste("CellRanger filltered cells:",ids[ind],sep=" ")) + xlab("Barcodes") + ylab("counts (UMI or Features)") +
    scale_x_continuous(trans = 'log2', breaks=xbreaks, labels = xlabels) +
    scale_y_continuous(trans = 'log2', breaks=ybreaks, labels = ylabels) +
    geom_line(aes(x=index, y=nCount_RNA, color=index<=cr_filtered_cells[ind] , group=1), size=1.75) +
    geom_line(aes(x=index, y=nFeature_RNA, color="Features", group=1), size=1.25)
  
  # barcode.umi.fig.list[[ids[ind]]] <- pl1
  
  return(pl1)
}


# plot_cellranger_cells takes as an argument the index of the sample in the list of ids
barcode.umi.fig.list <- lapply(1:length(ids), function(i){
  plot_cellranger_cells(i)
})

lapply(1:length(ids), function(i){
  barcode.umi.fig.list[[i]]
})

```


### Create the Seurat object

filter criteria: remove genes that do not occur in a minimum of 0 cells and remove cells that don't have a minimum of 300 features

<br/>

```{r create_seurat_object, warning=FALSE,error=FALSE, message=FALSE}
experiment.data <- do.call("cbind", d10x.data)

experiment.aggregate <- CreateSeuratObject(
  experiment.data,
  project = experiment.name,
  min.cells = 0,
  min.features = 300,
  names.field = 2,
  names.delim = "\\-")

# experiment.aggregate
# str(experiment.aggregate)
```

### The percentage of reads that map to the mitochondrial genome

* Low-quality / dying cells often exhibit extensive mitochondrial contamination.
* We calculate mitochondrial QC metrics with the PercentageFeatureSet function, which calculates the percentage of counts originating from a set of features.
* We use the set of all genes, in mouse these genes can be identified as those that begin with 'mt', in human data they begin with MT.

```{r calculate_pct_mito, warning=FALSE,error=FALSE, message=FALSE}
experiment.aggregate$percent.mito <- PercentageFeatureSet(experiment.aggregate, pattern = mt.marker)
summary(experiment.aggregate$percent.mito)
```

```{r calculate_rbc, warning=FALSE,error=FALSE, message=FALSE}
experiment.aggregate$percent.rbc <- PercentageFeatureSet(experiment.aggregate, pattern = hb.marker)
summary(experiment.aggregate$percent.rbc)
```

```{r calculate_pct_ribo, warning=FALSE,error=FALSE, message=FALSE}
experiment.aggregate$percent.ribo <- PercentageFeatureSet(experiment.aggregate, pattern = ribo.marker)
summary(experiment.aggregate$percent.ribo)
```

<!-- ### Lets spend a little time getting to know the Seurat object. -->

<br/>
<br/>

The Seurat object is the center of each single cell analysis. It stores __all__ information associated with the dataset, including data, annotations, analyses, etc. The R function slotNames can be used to view the slot names within an object.

```{r explore2, warning=FALSE,error=FALSE, message=FALSE}
slotNames(experiment.aggregate)
```

<!-- ```{r explore3, warning=FALSE,error=FALSE, message=FALSE} -->
<!-- head(experiment.aggregate[[]]) -->
<!-- ``` -->

<br/>

## Some basic QA/QC of the metadata, print tables of the 10% quantiles.

Show 10% quantiles for number of genes per cell per sample
```{r quantiles_1, warning=FALSE,error=FALSE,message=FALSE}
kable(do.call("cbind", tapply(experiment.aggregate$nFeature_RNA,
                      Idents(experiment.aggregate),quantile,probs=seq(0,1,0.1))),
      caption = "10% Quantiles of Genes/Cell by Sample") %>% kable_styling()
```

Show 10% quantiles for number of UMI per cell per sample
```{r quantiles_2, warning=FALSE,error=FALSE,message=FALSE}
kable(do.call("cbind", tapply(experiment.aggregate$nCount_RNA,
                                      Idents(experiment.aggregate),quantile,probs=seq(0,1,0.1))),
      caption = "10% Quantiles of UMI/Cell by Sample") %>% kable_styling()
```

Show 10% quantiles for number of mitochondrial percentage per cell per sample

```{r quantiles_3, warning=FALSE,error=FALSE,message=FALSE}
kable(round(do.call("cbind", tapply(experiment.aggregate$percent.mito, Idents(experiment.aggregate),quantile,probs=seq(0,1,0.1))), digits = 3),
      caption = "10% Quantiles of Percent Mitochondria by Sample") %>% kable_styling()
```

<br/>
<br/>

## Violin Plots

Violin plot of 1) number of genes, 2) number of UMI, 3) percent mitochondrial genes, 4) percent RBC markers, and 5) percent rRNA markers
```{r violins, warning=FALSE,error=FALSE,message=FALSE, fig.height=20}
part1.vln.plots <- VlnPlot(
  experiment.aggregate,
  features = c("nFeature_RNA", "nCount_RNA","percent.mito", "percent.rbc", "percent.ribo"),
  ncol = 1, pt.size = 0.3)

p1.vln.plot.filename <- "part1_vln_plots.png"
p1.vln.plot.path <- file.path(ggplot.directory, p1.vln.plot.filename)

part1.vln.plots
ggsave(p1.vln.plot.path, part1.vln.plots)
```

### Plot ridgeplots of the same data

```{r ridgeplot_pre, warning=FALSE,error=FALSE,message=FALSE}
RidgePlot(experiment.aggregate, features=c("nFeature_RNA","nCount_RNA", "percent.mito", "percent.rbc", "percent.ribo"), ncol = 2)
```

## Gene Plot, scatter plot of gene expression across cells (colored by sample).

<br/>

### Use these plots to determine the cutoffs for:

* max mitochondrial percentage
* min RNA count
* max RNA count
* min feature count
* max feature count
* max RBC
* min ribo


<br/>

```{r relationships, warning=FALSE,error=FALSE,message=FALSE}
FeatureScatter(experiment.aggregate, "nCount_RNA", "percent.mito")
FeatureScatter(experiment.aggregate, "nFeature_RNA", "percent.mito") 
FeatureScatter( experiment.aggregate, "nCount_RNA", "nFeature_RNA")
FeatureScatter( experiment.aggregate, "percent.ribo", "percent.rbc")
FeatureScatter( experiment.aggregate, "percent.ribo", "percent.mito")
FeatureScatter( experiment.aggregate, "percent.mito", "percent.rbc")
FeatureScatter( experiment.aggregate, "nFeature_RNA", "percent.rbc")
```

```{r convert_sparse_mat_to_mat, warning=FALSE,error=FALSE,message=FALSE }

# Function for converting sparse matrix to regular matrix.
# see https://programmerah.com/the-sparse-matrix-of-r-language-is-too-large-to-be-used-as-matrix-8856/

convert_sparse_mat_to_mat <- function(mat){
  tmp <- matrix(data=0L, nrow = mat@Dim[1], ncol = mat@Dim[2])

  row_pos <- mat@i+1
  col_pos <- findInterval(seq(mat@x)-1,mat@p[-1])+1
  val <- mat@x

  for (i in seq_along(val)){
    tmp[row_pos[i],col_pos[i]] <- val[i]
    }
  row.names(tmp) <- mat@Dimnames[[1]]
  colnames(tmp) <- mat@Dimnames[[2]]
  return(tmp)
}

```

## Sex determination

Violin plots showing expression of genes on the Y chromosome (male), along with 
plot of expression of XIST (X-inactive specific transcript) (female). 

```{r plot_y_and_xist, eval = FALSE}
Idents(experiment.aggregate) <- "orig.ident"

if (!file.exists(genes.file)) {
    suppressMessages(require(biomaRt))

    # initialize connection to mart, may take some time if the sites are
    # unresponsive.
    mart <- useMart("ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl")

    # fetch chromosome info plus some other annotations
    genes.table <- try(biomaRt::getBM(attributes = c("ensembl_gene_id", "external_gene_name",
        "description", "gene_biotype", "chromosome_name", "start_position"), mart = mart,
        useCache = F))

    if (!dir.exists("data/results")) {
        dir.create("data/results")
    }
    if (is.data.frame(genes.table)) {
        write.csv(genes.table, file = genes.file)
    }

    if (!file.exists(genes.file)) {
        download.file("https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/labs/misc/genes.table.csv",
            destfile = "data/results/genes.table.csv")
        genes.table = read.csv(genes.file)
    }

} else { 
  suppressMessages(require(rtracklayer)) 
  genes.table <- import(genes.file) 
}

genes.table <- genes.table[genes.table$gene_name %in% rownames(experiment.aggregate),]


chrY.genes <- subset(genes.table, seqnames(genes.table) == "chrY")
chrY.gene <- chrY.genes$gene_name

experiment.aggregate$pct_chrY = colSums(experiment.aggregate@assays$RNA@counts[chrY.gene,])/colSums(experiment.aggregate@assays$RNA@counts)

# FeatureScatter(experiment.aggregate, feature1 = xist.gene, feature2 = "pct_chrY")
sex.vln.plots <- VlnPlot(experiment.aggregate, features = c(xist.gene, "pct_chrY"))

sex.vln.plot.filename <- "part1_sex_vln_plots.png"
sex.vln.plot.path <- file.path(ggplot.directory, sex.vln.plot.filename)

ggsave(sex.vln.plot.path, sex.vln.plots)

sex.vln.plots

```

# Top 20 most highly expressed genes in this dataset

## Relative expression of each gene per cell
```{r relative expression, warning=FALSE,error=FALSE,message=FALSE}
rel.exp.bp.filename <- "part1_relative_expression_boxplot.png"
rel.exp.bp.path <- file.path(ggplot.directory, rel.exp.bp.filename)

par(mar = c(4, 8, 2, 1))
C <- experiment.aggregate[["RNA"]]$counts
C <- Matrix::t(Matrix::t(C)/Matrix::colSums(C)) * 100
C_mat <- convert_sparse_mat_to_mat(C)
# most_expressed <- order(apply(C, 1, median), decreasing = T)[20:1]
most_expressed <- order(apply(C_mat, 1, median), decreasing = T)[20:1]

# png(rel.exp.bp.path, width = 1620, height = 1125)
relative.expression.boxplot <- boxplot(as.matrix(t(C[most_expressed, ])), 
                                       cex = 0.8, 
                                       las = 1, 
                                       xlab = "% total count per cell", 
                                       col = (scales::hue_pal())(20)[20:1], 
                                       horizontal = TRUE)
# dev.off()

# relative.expression.boxplot

# relative.expression.boxplot
```

## Finally, save the original object and view the object.

Original dataset in Seurat class, with no filtering

```{r save, warning=FALSE,error=FALSE, message=FALSE}
saveRDS(experiment.aggregate,file=part1.rds.save.path)

save(list=c("sequencing.char.table",  
            "mapping.char.table.1",  
            "mapping.char.table.2",  
            "cell.char.table", 
            "barcode.umi.fig.list"), 
     file=part1.file.report.path)
```

## Session Information
```{r sessioinfo}
sessionInfo()
```

