---
output:
  html_document:
    pandoc_args: './common.yaml'
params:
  config.args: "./config.yaml"
---

```{r load_libraries, warning=FALSE,error=FALSE,message=FALSE, echo=FALSE}
# must have Seurat
library(Seurat)
library(kableExtra)
library(ggplot2)
library(yaml)
```


```{r user_vars, echo = FALSE}

# get the datetime and set it to variable for inline printing
todays.date <- Sys.Date()
formatted.date <- format(todays.date, format="%B %d %Y")

# Import the config file
# config.args <- read_yaml(params$config.args)
config.args <- read_yaml(params$config.args)

# set project variables
analyst.name <- config.args$project$`analyst-name`
formatted.samples <- config.args$project$samples
project.description <- config.args$project$`project-decription`
dataset.path <- config.args$data$parent_directory_path
r.obj.loc <- config.args$data$`rds-file-path` # Output folder

metrics.table.string <- config.args$analysis$part1$generate_metrics_tables
run_metrics_tables=TRUE

if (metrics.table.string == "FALSE"){
  run_metrics_tables=FALSE
}

part1.suffix <- config.args$analysis$part1$part1_suffix
part2.suffix <- config.args$analysis$part2$part2_suffix

#--------------------------------------------------#
# Set up the filenames by taking the normal output
# filename from the config and adding the suffix
#--------------------------------------------------#
part1.file.for.report <- config.args$analysis$part1$part1_report_tables_filename
part1.file.for.report <-sub(".RData", paste0(part1.suffix, ".RData"), part1.file.for.report)
part1.file.report.path <- file.path(r.obj.loc, part1.file.for.report)

part2.file.for.report <- config.args$analysis$part2$part2_report_tables_filename
part2.file.for.report <-sub(".RData", paste0(part2.suffix, ".RData"), part2.file.for.report)
part2.file.report.path <- file.path(r.obj.loc, part2.file.for.report)

# Load the report files
load(part1.file.report.path)
load(part2.file.report.path)

organism = config.args$project$organism

# setup the sample names for printing in the inline report
formatted.samples <- gsub("\"", "", formatted.samples)
formatted.samples <- gsub("\ ", "  \n ", formatted.samples)
formatted.samples <- gsub("-", " * ", formatted.samples)
formatted.samples <- paste0(formatted.samples, " <br /> ")

# generate list of ids for use in importing data
ids <- config.args$project$samples # sample names, need to match the folders containing the h5 files
ids <- gsub("-", "", ids)
ids <- gsub("\"", "", ids)
ids <- strsplit(ids, "[[:space:]]")
ids <- unlist(ids)


ggplot.directory.name <- config.args$analysis$part1$ggplot_dir
ggplot.directory <- file.path(r.obj.loc, ggplot.directory.name)
```

```{r get_plot_path, echo = FALSE}

get_png_plot_path <- function(filename){
  input.path <- file.path(ggplot.directory, filename)
  input.path
}

```

* This report was generated on `r formatted.date`
* organism: `r organism`

# Project Description:

`r project.description`

<br/>
<br/>

# Analysis software used
[Cellranger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger) was used to align reads and generate feature-barcode matrices.

[Seurat](http://satijalab.org/seurat/) (now Version 4) is a popular R package that is designed for QC, analysis, and exploration of single cell data. Seurat aims to enable users to identify and interpret sources of heterogeneity from single cell transcriptomic measurements, and to integrate diverse types of single cell data. Seurat was used to filter data, run pca, and generate and visualize clusters.

<br/>
<br/>

# Sample QC


### Sample names

`r formatted.samples `


<br/>
<br/>

`r if (run_metrics_tables) {"### Raw data metrics"}`
<!-- ### Raw data metrics -->

`r if (run_metrics_tables) {"Here is a table showing the raw sequencing metrics for all samples."}`
<br/>


```{r sequencing_characteristics, warning=FALSE,error=FALSE, message=FALSE, echo=FALSE, eval=run_metrics_tables}

sequencing.char.table
```

```{r mapping_characteristics_1, warning=FALSE,error=FALSE, message=FALSE, echo=FALSE, eval=run_metrics_tables}

mapping.char.table.1
```

```{r mapping_characteristics_2, warning=FALSE,error=FALSE, message=FALSE, echo=FALSE, eval=run_metrics_tables}

mapping.char.table.2
```

```{r cell_characteristics, warning=FALSE,error=FALSE, message=FALSE, echo=FALSE, eval=run_metrics_tables}

cell.char.table
```

<br/>
<br/>

### Basic first filter criteria applied

Remove genes that do not occur in a minimum of 0 cells and remove cells that don't have a minimum of 300 features


### Violin Plots of raw data
<br/>

##### nFeature: This is the number of different genes being represented by the data in each cell.
<br/>

##### nCount: This is the number of Unique Molecular Identifiers (UMIs) in each cell, which represents the number of unique RNA molecules in each cell (i.e. ignoring any pcr duplicates). 
<br/>

##### percent.mito: The percent of UMIs in each cell that are mitochondrial.
<br/>

##### percent.rbc: The percent of UMIs in each cell that are red blood cell marker genes.
<br/>

##### percent.ribo: The percent of UMIs in each cell that are ribosomal.
<br/>

```{r violins, warning=FALSE,error=FALSE,message=FALSE, fig.height=22, echo=FALSE}
part1.vln.plot.path <- file.path(ggplot.directory,
                                 paste0("part1", part1.suffix, "_vln_plots.png"))

knitr::include_graphics(part1.vln.plot.path)

```
<br/>
<br/>

### Gene plots of raw data

Scatter plots of gene expression across cells, colored by sample.
Samples will be filtered by the cutoffs listed. Cutoffs are also illustrated by lines on the gene plots.

```{r relationships, warning=FALSE,error=FALSE,message=FALSE, echo=FALSE}
# todo: fill out the filtering values below
# Set these cutoffs here and use them in the plots AND in the filtering
print("Cutoff values: ")
print(paste0("Remove cells with % mito greater than ", mito.cutoff))
print(paste0("Remove cells with more than ", feature.count.floor, " features (i.e. unique genes), and less than ", feature.count.ceiling))
print(paste0("Remove cells with less than ", RNA.count.floor, " counts (i.e. UMIs), or with more than ", RNA.count.ceiling))
print(paste0("Remove cells with less than ", ribo.cutoff, " percent of all counts coming from ribosomal markers."))
print(paste0("Remove cells with greater than ", rbc.cutoff, " percent of all counts coming from red blood cell markers."))
```
<br/>
<br/>

### nUMIs vs mitochondrial %

```{r ncount.mito.scatter, warning=FALSE,error=FALSE,message=FALSE, echo = FALSE}

part2.nCount.mito.path <- file.path(ggplot.directory,
                                 paste0("part2", part2.suffix, "_nCount_mito.png"))

knitr::include_graphics(part2.nCount.mito.path)

```

<br/>

### number of genes vs mitochondrial %

```{r nfeature.mito.scatter, warning=FALSE,error=FALSE,message=FALSE, echo = FALSE}
part2.nFeat.mito.path <- file.path(ggplot.directory,
                                 paste0("part2", part2.suffix, "_nFeat_mito.png"))

knitr::include_graphics(part2.nFeat.mito.path)
```

<br/>

### nUMIs vs number of genes

```{r ncount.nfeature.scatter, warning=FALSE,error=FALSE,message=FALSE, echo = FALSE}
part2.nCount.nFeat.path <- file.path(ggplot.directory,
                                 paste0("part2", part2.suffix, "_nCount_nFeat.png"))

knitr::include_graphics(part2.nCount.nFeat.path)
```

<br/>

### % UMIs from ribo genes vs % UMIs from RBCs

```{r ribo.rbc.scatter, warning=FALSE,error=FALSE,message=FALSE, echo = FALSE}
part2.ribo.rbc.path <- file.path(ggplot.directory,
                                 paste0("part2", part2.suffix, "_ribo_rbc.png"))

knitr::include_graphics(part2.ribo.rbc.path)
```

<br/>

### % UMIs from ribo genes vs mitochondrial %

```{r ribo.mito.scatter, warning=FALSE,error=FALSE,message=FALSE, echo = FALSE}
part2.ribo.mito.path <- file.path(ggplot.directory,
                                 paste0("part2", part2.suffix, "_ribo_mito.png"))

knitr::include_graphics(part2.ribo.mito.path)
```

<br/>

### mitochondrial % vs % UMIs from RBCs
```{r mito.rbc.scatter, warning=FALSE,error=FALSE,message=FALSE, echo = FALSE}
part2.mito.rbc.path <- file.path(ggplot.directory,
                                 paste0("part2", part2.suffix, "_mito_rbc.png"))

knitr::include_graphics(part2.mito.rbc.path)
```

<br/>

### Number of genes vs % UMIs from RBCs

```{r nfeature.rbc.scatter, warning=FALSE,error=FALSE,message=FALSE, echo = FALSE}
part2.nFeat.rbc.path <- file.path(ggplot.directory,
                                 paste0("part2", part2.suffix, "_nFeat_rbc.png"))

knitr::include_graphics(part2.nFeat.rbc.path)
```

<br/>

#### Tables showing how many cells remain of each sample after the above filtering steps.
<br/>

```{r apply_first_filtering, warning=FALSE,error=FALSE,message=FALSE, echo=FALSE}
pretty.combo.table
```
<br/>  
<br/>  

### Ridgeplots of raw data

Ridgeplots of the same data as the violin plots above

```{r ridgeplot_pre, warning=FALSE,error=FALSE,message=FALSE, echo=FALSE}
part2.ridgeplot.prefiltering.path <- file.path(ggplot.directory,
                                 paste0("part2", part2.suffix, "_ridgeplot_prefiltering.png"))

knitr::include_graphics(part2.ridgeplot.prefiltering.path)
```
<br/>

### Ridgeplots after filtering
<br/>

```{r ridgeplot_first_post, warning=FALSE,error=FALSE,message=FALSE, echo = FALSE}
part2.ridgeplot.postfiltering.path <- file.path(ggplot.directory,
                                 paste0("part2", part2.suffix, "_ridgeplot_postfiltering.png"))

knitr::include_graphics(part2.ridgeplot.postfiltering.path)
```
<br/>
<br/>

### Session Information
```{r sessioinfo}
sessionInfo()
```
