---
title: "scRNA-seq QC Report"
output:
  pdf_document:
    toc: true
    number_sections: true
    fig_caption: true
    latex_engine: xelatex
    includes:
      in_header: common-pdf.tex
params:
  config.args: "./config.yaml"
---



```{r setup, include=FALSE}
# install.packages("tinytex")
# tinytex::install_tinytex()   # downloads + sets up ~/.TinyTeX
# tinytex::is_tinytex()        # should return TRUE

knitr::opts_chunk$set(
  echo = FALSE, warning=FALSE,error=FALSE, message=FALSE
)

# source helpers that sit beside this Rmd
source("functions_for_report.R")
```



```{r import_libraries, eval=TRUE}
if (!suppressMessages(require(Seurat))) stop("Seurat is required")
if (!suppressMessages(require(ggplot2))) stop("ggplot2 is required")
if (!suppressMessages(require(yaml))) stop("yaml is required")
if (!suppressMessages(require(knitr))) stop("knitr is required")
if (!suppressMessages(require(kableExtra))) stop("kableExtra is required")
if (!suppressMessages(require(dplyr))) stop("dplyr is required")
if (!suppressMessages(require(scales))) stop("scales is required")

```


```{r user_vars, eval=TRUE}

# get the datetime and set it to variable for inline printing
todays.date <- Sys.Date()
formatted.date <- format(todays.date, format="%B %d %Y")

# Import the config file
config.args <- read_yaml(params$config.args)

run_metrics_tables=TRUE
is_tenx=TRUE

metrics.table.string <- config.args$analysis$part1$generate_metrics_tables

if (metrics.table.string == "FALSE"){
  run_metrics_tables=FALSE
}


################################################################################
# set project variables
################################################################################
analyst.name <- config.args$project$`analyst-name`
formatted.samples <- config.args$project$samples # Sanitized below
experiment.name = config.args$project$`project-name` 
project.description <- config.args$project$`project-decription`
dataset.path <- config.args$data$parent_directory_path
import.h5 <- config.args$data$`import-h5` # set to false if you want to import the matrix files
organism = config.args$project$organism


################################################################################
# Set output directory paths
################################################################################

r.obj.loc <- config.args$data$`rds-file-path` # Output folder

# When trying different parameters in the analysis you can create
# alternate output files with this 'suffix' argument in the config
part1.suffix <- config.args$analysis$part1$part1_suffix
unfiltered.seurat.obj.file <- config.args$analysis$part1$part1_rds_save_filename
unfiltered.seurat.obj.file <- sub(".RDS", paste0(part1.suffix, ".RDS"), unfiltered.seurat.obj.file)
unfiltered.seurat.obj.path <- file.path(r.obj.loc, unfiltered.seurat.obj.file)
QC.tables.file <- config.args$analysis$part1$part1_report_tables_filename
QC.tables.file <-sub(".RData", paste0(part1.suffix, ".RData"), QC.tables.file)
QC.tables.path <-file.path(r.obj.loc, QC.tables.file)

ggplot.directory.name <- config.args$analysis$part1$ggplot_dir
ggplot.directory <- file.path(r.obj.loc, ggplot.directory.name)

# create ggplot folder if it doesn't exist already
if (!file.exists(ggplot.directory)) {
 dir.create(ggplot.directory, showWarnings = FALSE, recursive = TRUE)
}

# create rds folder if it doesn't exist already
if (!file.exists(r.obj.loc)) {

 dir.create(r.obj.loc, showWarnings = FALSE, recursive = TRUE)

}



################################################################################
# Set gene subset markers (mito genes, ribo genes, etc)
################################################################################

if (organism == "human"){
  mt.marker = "^MT-"
  hb.marker = "^HB[^(P)]"
  ribo.marker = "^RP[SL]"
  xist.gene = "XIST"
  y.genes <- readLines("y_genes_hg38_symbols.txt")
  genes.file = "/projects/b1012/xvault/REFERENCES/10xGenomics/RNA-seq/hg38/refdata-gex-GRCh38-2020-A/genes/genes.gtf"
} else if (organism == "mouse") {
  mt.marker = "^mt-"
  hb.marker = "^Hb[^(p)]"
  ribo.marker = "^Rp[sl]" 
  y.genes <- readLines("y_genes_mm10_symbols.txt")
  genes.file = "/projects/b1012/xvault/REFERENCES/10xGenomics/RNA-seq/mm10/refdata-gex-mm10-2020-A/genes/genes.gtf"
  xist.gene = "Xist"
} else {
  mt.marker = "NONE"
  hb.marker = "NONE"
  ribo.marker = "NONE"
  xist.gene = "NONE"
}


################################################################################
# Format the sample names
################################################################################

ids <- sanitize_sample_names(formatted.samples)

```


## Report and variable details

This report was generated on `r formatted.date`

* organism: `r organism`

* starting with raw data at \nolinkurl{`r dataset.path`}
* saving rds object to \nolinkurl{`r unfiltered.seurat.obj.path`}
* saving file for report at \nolinkurl{`r QC.tables.path`}
* Importing h5?: `r import.h5`

## Project Description:

`r project.description`


## Sample Names

`r formatted.samples `


# Part 1: Loading data from CellRanger into R

## Sequencing QC table

For more insight on what the different metrics presented here mean, see this article on the 10x website: https://www.10xgenomics.com/analysis-guides/quality-assessment-using-the-cell-ranger-web-summary

<br/>
<br/>

```{r prep_qc_table, eval=run_metrics_tables, message=FALSE, warning=FALSE}
# --- Import the metrics summaries (quiet) ---
qc.table <- make_qc_table(ids)
render_formatted_table(qc.table)
```



## Load the Cell Ranger Matrix Data and create the base Seurat object.

<!-- Import the 10x data (either matrices or h5) and create a merged seurat object with counts of each sample as layers. -->

Objects created for:



```{r make_seurat_objs}

# make a Seurat object per sample
seurat.list <- lapply(ids, function(s) {
  # m <- read_10x_any(s)
  m <- read_10x_any(s)
  s_fixed <- gsub("-", "_", s)
  colnames(m) <- paste0(sub("-.*$", "", colnames(m)), "-", s_fixed)  # keep 10x barcode-left, add sample suffix
  o <- CreateSeuratObject(counts = m, project = experiment.name, min.features = 300)
  o$sample_id <- s_fixed
  print(paste0(unique(s_fixed)))
  o
})
names(seurat.list) <- gsub("-", "_", ids)
```

```{r set_percent_feats}

for (i in 1:length(seurat.list)) {
  seurat.list[[i]][["percent.mt"]] <- PercentageFeatureSet(seurat.list[[i]], pattern = mt.marker)
  # Calculate percentage of counts from hemoglobin genes (RBC markers)
  seurat.list[[i]][["percent.hb"]] <- PercentageFeatureSet(seurat.list[[i]], pattern = hb.marker)
  seurat.list[[i]][["percent.y"]] <- PercentageFeatureSet(seurat.list[[i]], features = y.genes)
  seurat.list[[i]][["percent.ribo"]] <- PercentageFeatureSet(seurat.list[[i]], pattern = ribo.marker)
  
}

```

```{r gather_qc_data_for_tables}
# Alphabetical sorting (default)
nCount_data       <- make_metric_table_from_list(seurat.list, "nCount_RNA")
nFeature_data     <- make_metric_table_from_list(seurat.list, "nFeature_RNA")
percent_mt_data   <- make_metric_table_from_list(seurat.list, "percent.mt")
percent_hb_data   <- make_metric_table_from_list(seurat.list, "percent.hb")
percent_y_data    <- make_metric_table_from_list(seurat.list, "percent.y")
percent_ribo_data <- make_metric_table_from_list(seurat.list, "percent.ribo")

# Or with a custom sample order
# sample_order <- c("sample_1", "sample_2",
#                   "sample_3", "sample_4")
# 
# nCount_data <- make_metric_table_from_list(seurat.list, "nCount_RNA", sample_levels = sample_order)



to.save.for.report <- c("nCount_data",
                        "nFeature_data", 
                        "percent_mt_data",
                        "percent_hb_data",
                        "percent_y_data", 
                        "percent_ribo_data")

save(list = to.save.for.report, file=QC.tables.path)
```

# Part 2: QC plots

## nCount RNA distribution

```{r p_nCount_plot}
p_nCount <- vln_boxplot(data = nCount_data,  
                          x = Sample, 
                          y = nCount_RNA,
                          title = "nCount_RNA Distribution per Sample")
# p_nCount <- ggplot(nCount_data, aes(x = Sample, y = nCount_RNA)) +
#   geom_violin() +
#   geom_boxplot(width = 0.1) +
#   theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
#   ggtitle("nCount_RNA Distribution per Sample")

p_nCount

save_png_plot(p_nCount, "nCount_RNA_distribution.png")
```


## nFeature distribution

```{r p_nFeature_plot}

p_nFeature <- vln_boxplot(data = nFeature_data,  
                          x = Sample, 
                          y = nFeature_RNA,
                          title = "nFeature_RNA Distribution per Sample")
  


p_nFeature
save_png_plot(p_nFeature, "nFeature_RNA_distribution.png")
```

## Percent mito distribution

```{r pct_mito_plot}

p_percent_mt <- vln_boxplot(data = percent_mt_data,  
                          x = Sample, 
                          y = percent.mt,
                          title = "Percent mitochondrial content per Sample")


p_percent_mt
save_png_plot(p_percent_mt, "percent_mt_distribution.png")
```

## Percent RBC distribution

```{r pct_hb_plot}

p_percent_hb <- vln_boxplot(data = percent_hb_data,  
                          x = Sample, 
                          y = percent.hb,
                          title = "Percent hemoglobin content (RBCs) per Sample")

p_percent_hb
save_png_plot(p_percent_hb, "percent_hb_distribution.png")

```

## Percent ribo genes distribution

```{r pct_ribo_plot}

p_percent_ribo <- vln_boxplot(data = percent_ribo_data,  
                          x = Sample, 
                          y = percent.ribo,
                          title = "Percent ribo gene expression per Sample")

p_percent_ribo
save_png_plot(p_percent_ribo, "percent_ribo_distribution.png")

```


## Percent Y-chr genes distribution

```{r pct_y_plot}

p_percent_y <- vln_boxplot(data = percent_y_data,  
                          x = Sample, 
                          y = percent.y,
                          title = "Percent Y chr gene expression per Sample")

p_percent_y
save_png_plot(p_percent_y, "percent_y_distribution.png")
```

```{r merge_list_into_agg_obj}
# merge => v5 stores each sample as its own layer(s) inside RNA
# seurat.obj.v5 <- Reduce(function(x, y) merge(x, y, add.cell.ids = c(unique(x$sample_id), unique(y$sample_id))), seurat.list)

seurat.obj.v5 <- merge(
  x            = seurat.list[[1]],
  y            = seurat.list[-1],
  add.cell.ids = ids,
  project      = "MBE"
)

seurat.obj.v5 <- JoinLayers(seurat.obj.v5)

# The layer names are now defaults, e.g. counts.1, counts.2, etc. Rename the layers
# with sample names

# 1) make sure your per-cell sample label is exactly what you want in the layer names
#    (factor labels become the suffix after 'counts.')
seurat.obj.v5$sample_id <- factor(
  seurat.obj.v5$sample_id,
  levels = unique(seurat.obj.v5$sample_id),                  # or your custom order
  labels = make.names(unique(seurat.obj.v5$sample_id))       # sanitize: replace -/spaces with _
)

# 2) split the assay into layers by that factor
seurat.obj.v5[["RNA"]] <- split(seurat.obj.v5[["RNA"]], f = seurat.obj.v5$sample_id)

```


```{r save_obj}
saveRDS(seurat.obj.v5, unfiltered.seurat.obj.path)
```



## Session Information
```{r sessioinfo, eval=TRUE}
sessionInfo()
```

