---
title: "scRNA-seq QC Report"
output:
  pdf_document:
    toc: true
    number_sections: true
    fig_caption: true
    latex_engine: xelatex
params:
  config.args: "./config.yaml"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  warning=FALSE,error=FALSE, message=FALSE, echo=FALSE
)

source("functions_for_report.R")
```

# Part 2: Some QA/QC, filtering, normalization, and assigning variable genes

```{r libraries, warning=FALSE,error=FALSE,message=FALSE}
library(Seurat)
# library(biomaRt)
library(ggplot2)
# library(knitr)
# library(kableExtra)
# library(future)
library(yaml)
```

```{r prepare_inline_variables, echo = FALSE}
# get the datetime and set it to variable
todays.date <- Sys.Date()
formatted.date <- format(todays.date, format="%B %d %Y")

config.args <- read_yaml(params$config.args)

```

```{r user_vals, warning=FALSE,error=FALSE,message=FALSE, echo=FALSE}
r.obj.loc <- config.args$data$`rds-file-path`
organism <- config.args$project$organism

################################################################################
# QC cutoffs
################################################################################

mito.ceiling = config.args$analysis$part2$mito_ceiling
RNA.count.floor = config.args$analysis$part2$RNA_count_floor
RNA.count.ceiling = config.args$analysis$part2$RNA_count_ceiling
feature.count.floor = config.args$analysis$part2$feature_count_floor
feature.count.ceiling = config.args$analysis$part2$feature_count_ceiling
rbc.ceiling = config.args$analysis$part2$rbc_ceiling
ribo.floor = config.args$analysis$part2$ribo_floor

node.type <- config.args$analysis$node_type


                filename_seurat_obj_merged: "_2_seurat5_obj_merged.RDS"
                filename_seurat_obj_integrated: "_3_seurat5_obj_integrated.RDS"
################################################################################
# I/O paths
################################################################################
part1.suffix <- config.args$analysis$part1$part1_suffix
part2.suffix <- config.args$analysis$part2$part2_suffix

unfiltered.seurat.obj.file <- config.args$analysis$part2$unfiltered.seurat.obj_filename # Get the absolute path
unfiltered.seurat.obj.file <- sub(".RDS", paste0(part1.suffix, ".RDS"), unfiltered.seurat.obj.file ) # add the suffix
unfiltered.seurat.obj.path <- file.path(r.obj.loc, unfiltered.seurat.obj.file) # make the full path

merged.seurat.obj.file <- config.args$analysis$part2$merged.seurat.obj.file
merged.seurat.obj.file <- sub(".RDS", paste0(part2.suffix, ".RDS"), merged.seurat.obj.file )
merged.seurat.obj.path <- file.path(r.obj.loc, merged.seurat.obj.file)

integrated.seurat.obj.file <- config.args$analysis$part2$integrated.seurat.obj.file
integrated.seurat.obj.file <- sub(".RDS", paste0(part2.suffix, ".RDS"), integrated.seurat.obj.file )
integrated.seurat.obj.path <- file.path(r.obj.loc, integrated.seurat.obj.file)

# part2.file.for.report <- config.args$analysis$part2$part2_report_tables_filename
# part2.file.for.report <-sub(".RData", paste0(part2.suffix, ".RData"), part2.file.for.report)
# part2.file.report.path <- file.path(r.obj.loc, part2.file.for.report)

QC.tables.file <- config.args$analysis$part1$part1_report_tables_filename
QC.tables.file <-sub(".RData", paste0(part1.suffix, ".RData"), QC.tables.file)
QC.tables.path <-file.path(r.obj.loc, QC.tables.file)

ggplot.directory.name <- config.args$analysis$part1$ggplot_dir
ggplot.directory <- file.path(r.obj.loc, ggplot.directory.name)
```


## Report and variable details

This report was generated on `r formatted.date`

* node.type: `r node.type`

* starting with rds object at `r unfiltered.seurat.obj.path`
* saving rds object to `r part2.rds.save.path`
* saving file for report at `r part2.file.report.path`

* mito ceiling: `r mito.ceiling`
* RNA count floor: `r RNA.count.floor`
* RNA count ceiling: `r RNA.count.ceiling`
* Feature count floor: `r feature.count.floor`
* Feature count ceiling: `r feature.count.ceiling`
* RBC ceiling: `r rbc.ceiling`
* Ribosomal floor: `r ribo.floor`



```{r get_num_cores_available, echo = FALSE}
if (node.type == "compute"){
  # If using a compute node, then put this in the submission script:
  # export MC_CORES=${SLURM_NTASKS}
  num.cores.available <- as.numeric(Sys.getenv("SLURM_NTASKS"))
} else {
  library(parallel)
  num.cores.available <- detectCores() 
}
```

## 

```{r load_data, warning=FALSE,error=FALSE,message=FALSE}
seurat5.obj <- readRDS(file=unfiltered.seurat.obj.path)
set.seed(12345)
```


<br/>

# QC plots

Just as in part 1, but shaded to show the filtering cutoffs

```{r load_QC_data}
# Same plots as part 1, with a red line showing the cutoffs
load(QC.tables.path)
```

## nCount RNA distribution

```{r shaded_nCount_plot}
p_nCount <- shaded_vln_boxplot(data = nCount_data,  
                          x = Sample, 
                          y = nCount_RNA,
                          title = "nCount_RNA Distribution per Sample",
                          floor = RNA.count.floor,
                          ceiling = RNA.count.ceiling)

p_nCount 
save_png_plot(p_nCount, "nCount_RNA_distribution_shaded.pdf")
```
## nFeature distribution

```{r p_nFeature_plot}

p_nFeature <- shaded_vln_boxplot(data = nFeature_data,  
                          x = Sample, 
                          y = nFeature_RNA,
                          title = "nFeature_RNA Distribution per Sample",
                          floor = feature.count.floor,
                          ceiling = feature.count.ceiling)
  


p_nFeature
save_png_plot(p_nFeature, "nFeature_RNA_distribution_shaded.pdf")
```

## Percent mito distribution

```{r pct_mito_plot}

p_percent_mt <- shaded_vln_boxplot(data = percent_mt_data,  
                          x = Sample, 
                          y = percent.mt,
                          title = "Percent mitochondrial content per Sample",
                          ceiling = mito.ceiling)


p_percent_mt
save_png_plot(p_percent_mt, "percent_mt_distribution.pdf")
```

## Percent RBC distribution

```{r pct_hb_plot}

p_percent_hb <- shaded_vln_boxplot(data = percent_hb_data,  
                          x = Sample, 
                          y = percent.hb,
                          title = "Percent hemoglobin content (RBCs) per Sample",
                          ceiling = rbc.ceiling)

p_percent_hb
save_png_plot(p_percent_hb, "percent_hb_distribution_shaded.pdf")

```

## Cell filtering stats

We use the cutoffs above to filter out cells. 

```{r apply_filtering, warning=FALSE,error=FALSE,message=FALSE}
before.table <- table(seurat5.obj$sample_id)

seurat5.obj <- subset(seurat5.obj, percent.mt <= mito.ceiling)
                                 # nCount_RNA >= RNA.count.floor & 
                                 # nCount_RNA <= RNA.count.ceiling &
                                 # nFeature_RNA >=feature.count.floor & 
                                 # nFeature_RNA <= feature.count.ceiling & 
                                 # percent.hb <= rbc.ceiling)
# seurat5.obj <- subset(seurat5.obj, nCount_RNA >= RNA.count.floor & nCount_RNA <= RNA.count.ceiling)
# seurat5.obj <- subset(seurat5.obj, nFeature_RNA >= feature.count.floor & nFeature_RNA <= feature.count.ceiling)
# seurat5.obj <- subset(seurat5.obj, percent.rbc >= rbc.cutoff)
# seurat5.obj <- subset(seurat5.obj, percent.ribo <= ribo.cutoff)

after.table <- table(seurat5.obj$sample_id)

combo.table <- rbind(before.table,
                     after.table)

row.names(combo.table) <- c("# cells before filtering", "# cells after filtering")

combo.table
```

# Analysis without integration

SCTransform, PCA and UMAP generation

```{r process_merged_pbject}

seurat5.obj <- SCTransform(seurat5.obj)
seurat5.obj <- RunPCA(seurat5.obj)
seurat5.obj <- RunUMAP(seurat5.obj, dims = 1:30)

print(paste0("Saving merged object to ", merged.seurat.obj.path))
saveRDS(seurat5.obj, merged.seurat.obj.path)
```

## Umap of the Seurat object before integration

```{r merged_obj_umap}
merged.dimplot <- DimPlot(seurat5.obj, reduction = "umap", group.by = "sample_id")

save_png_plot(merged.dimplot, "umap_merged_seruat_obj.png")
merged.dimplot
```

# Integration

```{r integrate_seurat_obj}
seurat5.obj.integrated <- IntegrateLayers(object = seurat5.obj, 
                                          method = CCAIntegration, 
                                          normalization.method = "SCT", 
                                          verbose = F)

saveRDS(seurat5.obj.integrated)

```


```{r clustering, eval=FALSE}

seurat5.obj <- FindNeighbors(seurat5.obj, reduction = "integrated.dr", dims = 1:30)
seurat5.obj <- FindClusters(seurat5.obj, resolution = 0.6)



seurat5.obj <- RunUMAP(seurat5.obj, dims = 1:30, reduction = "integrated.dr")
DimPlot(seurat5.obj, reduction = "umap", group.by = "sample_id")

# perform differential expression
seurat5.obj <- PrepSCTFindMarkers(seurat5.obj)
Idents(seurat5.obj)

# Runs MUCH faster with presto package
all.markers <- FindAllMarkers(seurat5.obj, verbose = TRUE)
```



```{r save_r, eval=FALSE}
saveRDS(seurat5.obj, file=part2.rds.save.path)

to.save.for.report <- c("mito.cutoff", 
                        "RNA.count.floor", 
                        "RNA.count.ceiling", 
                        "feature.count.floor", 
                        "feature.count.ceiling", 
                        "rbc.cutoff", 
                        "ribo.cutoff",
												"pretty.combo.table"
                        )

save(list = to.save.for.report, file=part2.file.report.path)
```

## Session Information
```{r session_info, warning=FALSE,error=FALSE, message=FALSE}
sessionInfo()
```

