---
output:
  html_document:
    pandoc_args: './common.yaml'
params:
  config.args: "./config.yaml"
---

# Part 3: PCA and choice in number of PCS

## Load libraries
```{r load_libraries, warning=FALSE, error=FALSE, message=FALSE}
library(Seurat)
library(ggplot2)
library(kableExtra)
# library(future)
# library(doMC)
library(dplyr)
library(yaml)
```

```{r prepare_inline_variables, echo = FALSE}
# get the datetime and set it to variable
todays.date <- Sys.Date()
formatted.date <- format(todays.date, format="%B %d %Y")

config.args <- read_yaml(params$config.args)
```

```{r user_vals, warning=FALSE, error=FALSE, message=FALSE, echo=FALSE}
r.obj.loc <- config.args$data$`rds-file-path`

run_tests <- config.args$analysis$run_tests
node.type <- config.args$analysis$node_type

organism <- config.args$project$organism
cell.cycle.string <- config.args$project$run_cell_cycle
run_cell_cycle=TRUE

if (cell.cycle.string == "FALSE"){
  run_cell_cycle=FALSE
} 

batch.correction.method <- config.args$analysis$part3$batch_correction

######################################################################
# Default will be no integration of the samples if there is only 1
######################################################################
ids <- config.args$project$samples # sample names, need to match the folders containing the h5 files
ids <- gsub("-", "", ids)
ids <- gsub("\"", "", ids)
ids <- strsplit(ids, "[[:space:]]")
ids <- unlist(ids)

num.samples <- length(ids)


run.scale.data = TRUE

if (num.samples > 1){
  run.integrate.data = TRUE
} else {
  run.integrate.data = FALSE
}

run.sctransform = FALSE

regression.vars.1 <- c("percent.mito", "nFeature_RNA")
cell.cycle.vars <- c("S.Score", "G2M.Score") 

if (batch.correction.method %in% c("orig", "orig_ident", "orig.ident")){
  append(regression.vars, "orig.ident")
} else if (batch.correction.method == "sctransform"){
  run.sctransform = TRUE
}

# generate list of ids for use in importing data
part2.suffix <- config.args$analysis$part2$part2_suffix
part3.suffix <- config.args$analysis$part3$part3_suffix

part3.rds.open.file <- config.args$analysis$part3$part3_rds_open_filename
part3.rds.open.file <- sub(".RDS", paste0(part2.suffix, ".RDS"), part3.rds.open.file )
part3.rds.open.path <- file.path(r.obj.loc, part3.rds.open.file)

part3.rds.save.file <- config.args$analysis$part3$part3_rds_save_filename
part3.rds.save.file <- sub(".RDS", paste0(part3.suffix, ".RDS"), part3.rds.save.file )
part3.rds.save.path <- file.path(r.obj.loc, part3.rds.save.file)

part3.report.file <- config.args$analysis$part3$part3_report_tables_filename
part3.report.path <- file.path(r.obj.loc, part3.report.file)

ggplot.directory.name <- config.args$analysis$part1$ggplot_dir
ggplot.directory <- file.path(r.obj.loc, ggplot.directory.name)

```

```{r define_save_plots, echo = FALSE}

save_png_plot <- function(plot, filename){
  save.path <- file.path(ggplot.directory, paste0("part3", part3.suffix, "_", filename))
  ggsave(save.path, plot)
}

```

### This report was generated on `r formatted.date`

* node.type: `r node.type`
* batch correction method: `r batch.correction.method`

* starting with rds object at `r part3.rds.open.path`
* saving rds object to `r part3.rds.save.path`
* saving file for report at `r part3.report.path`


```{r get_n.cores, echo = FALSE}
if (node.type == "compute"){
  # If using a compute node, then put this in the submission script:
  # export MC_CORES=${SLURM_NTASKS}
  n.cores <- as.numeric(Sys.getenv("SLURM_NTASKS")) - 1
} else {
  library(parallel) 
  n.cores <- detectCores() - 1
}
```

## Load the Seurat object from part 2

```{r load_data, warning=FALSE, error=FALSE, message=FALSE}
experiment.aggregate <- readRDS(file=part3.rds.open.path)
```


## Next we want to normalize the data

After filtering out cells from the dataset, the next step is to normalize the data. By default, we employ a global-scaling normalization method LogNormalize that normalizes the gene expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and then log-transforms the data.

```{r normalize_help, warning=FALSE, error=FALSE, message=FALSE, eval=FALSE}
?NormalizeData
```

### Scale Data and Find Variable Genes

The function FindVariableFeatures identifies the most highly variable genes (default 2000 genes) by fitting a line to the relationship of log(variance) and log(mean) using loess smoothing, uses this information to standardize the data, then calculates the variance of the standardized data.  This helps avoid selecting genes that only appear variable due to their expression level.

However, instead of using the variable genes function, lets instead assign variable genes to a set of "minimally expressed" genes.

```{r normalize_find_var_genes, warning=FALSE, error=FALSE, message=FALSE}
# split the dataset into a list of seurat objects
experiment.split <- SplitObject(experiment.aggregate, split.by = "orig.ident")

# normalize and identify variable features for each dataset independently
experiment.split <- lapply(X = experiment.split, FUN = function(single.seurat.object) { 
  if (!run.sctransform){
    print(paste0("Running normalize and scaledata on ", unique(single.seurat.object$orig.ident)))
    single.seurat.object <- NormalizeData(single.seurat.object) 
    min.value = 1  
    min.cells = 10  
    num.cells <- Matrix::rowSums(GetAssayData(single.seurat.object, slot = "count") > min.value)  
    genes.use <- names(num.cells[which(num.cells >= min.cells)])  
    length(genes.use)   
    VariableFeatures(single.seurat.object) <- genes.use 
    single.seurat.object <- ScaleData(single.seurat.object, vars.to.regress = regression.vars.1)
    single.seurat.object
  } else {
    print(paste0("Running sctransform on ", unique(single.seurat.object$orig.ident)))
    single.seurat.object <- SCTransform(single.seurat.object, 
                                        vars.to.regress = regression.vars.1)
    
    if (num.samples == 1){
      print("Throwing single.seurat.object into experiment.aggregate")
      experiment.aggregate <- single.seurat.object
    } else {  
      single.seurat.object
    }
  }
})
```


```{r sample_integration, warning=FALSE, error=FALSE, message=FALSE, eval = run.integrate.data, echo = run.integrate.data}

if (run.sctransform){
  print("running integration on sctransform samples")
  features <- SelectIntegrationFeatures(object.list = experiment.split, nfeatures = 3000)
  experiment.split <- PrepSCTIntegration(object.list = experiment.split, anchor.features = features)

  immune.anchors <- FindIntegrationAnchors(object.list = experiment.split, 
                                           normalization.method = "SCT", 
                                           anchor.features = features)
  
  experiment.aggregate <- IntegrateData(anchorset = immune.anchors, 
                                       normalization.method = "SCT")
} else {
  print("Running integration on normalized and scaled data (NOT sctransform)")
  features <- SelectIntegrationFeatures(object.list = experiment.split)
  anchors <- FindIntegrationAnchors(object.list = experiment.split, anchor.features = features)

  experiment.aggregate <- IntegrateData(anchorset = anchors)
}
```


`r if (run_cell_cycle) {"### Calculate Cell-Cycle with Seurat"}`
`r if (run_cell_cycle) {"[Dissecting the multicellular ecosystem of metastatic melanoma by single-cell RNA-seq](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4944528/)"}`

`r if (run_cell_cycle) {"The list of genes comes with Seurat (only for human). We can use biomaRt to convert the human gene symbols to mouse."}`



```{r cellcycle, warning=FALSE, error=FALSE, eval = run_cell_cycle}
# DefaultAssay(experiment.aggregate) <- "RNA" # https://github.com/satijalab/seurat/issues/1717

# Mouse Code
convertHumanGeneList <- function(x){
  require("biomaRt")
  human <- useMart("ensembl", dataset = "hsapiens_gene_ensembl", host = "https://apr2022.archive.ensembl.org/") 
  mouse <- useMart("ensembl", dataset = "mmusculus_gene_ensembl", host = "https://apr2022.archive.ensembl.org/")

  genes = getLDS(attributes = c("hgnc_symbol"), filters = "hgnc_symbol", values = x, mart = human, attributesL = c("mgi_symbol"), martL = mouse, uniqueRows=T)

  humanx <- unique(genes[, 2])

  # Print the first 6 genes found to the screen
  print(head(humanx))
  return(humanx)
}

if (organism == "mouse"){
  tryCatch(
    {
      m.s.genes <- convertHumanGeneList(cc.genes.updated.2019$s.genes)  
      m.g2m.genes <- convertHumanGeneList(cc.genes.updated.2019$g2m.genes)
    }, error=function(cond) {
            message("Error in Biomart lookup for converting cell cycle genes.")
            message("Using the saved versions instead.")
            message("Here's the original error message:")
            message(cond)
            message("")
    })
  
  # I tried loading this in the exception handling above, but it wasn't working
  if (!exists("m.s.genes")){
    load(file="mouse_cell_cycle_genes.RData")
  }

  # Create our Seurat object and complete the initialization steps
  experiment.aggregate <- CellCycleScoring(experiment.aggregate, s.features = m.s.genes, g2m.features = m.g2m.genes, set.ident = TRUE)
  
} else if (organism == "human"){
  # Human Code
  s.genes <- (cc.genes$s.genes)
  g2m.genes <- (cc.genes$g2m.genes)

  # Create our Seurat object and complete the initialization steps
  experiment.aggregate <- CellCycleScoring(experiment.aggregate, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
}
```

#### Table of cell cycle (seurate)

```{r cellcycle_table, warning=FALSE, error=FALSE, message=FALSE, eval = run_cell_cycle}
cell.cycle.table <- table(experiment.aggregate@meta.data$Phase) %>% kable(caption = "Number of Cells in each Cell Cycle Stage", col.names = c("Stage", "Count"), align = "c") %>% kable_styling()

cell.cycle.table
```

#### Fixing the defualt "Ident" in Seurat

```{r update_ident, warning=FALSE, error=FALSE, message=FALSE}
table(Idents(experiment.aggregate))
## So lets change it back to samplename
Idents(experiment.aggregate) <- "orig.ident"
table(Idents(experiment.aggregate))
```


### Scale Data

ScaleData - Scales and centers genes in the dataset. If variables are provided in vars.to.regress, they are individually regressed against each gene, and the resulting residuals are then scaled and centered unless otherwise specified. Here we regress out cell cycle results S.Score and G2M.Score, percentage mitochondria (percent.mito) and the number of features (nFeature_RNA).


```{r scale_data_cell_cycle, warning=FALSE, error=FALSE, message=FALSE, eval=run_cell_cycle}


experiment.aggregate <- ScaleData(experiment.aggregate, 
                                  vars.to.regress = cell.cycle.vars,
                                  do.scale = FALSE)
```


## Dimensionality reduction with PCA

Next we perform PCA (principal components analysis) on the scaled data.  

```{r pca_help, warning=FALSE, error=FALSE, message=FALSE, eval=FALSE, echo = FALSE}
?RunPCA
```

```{r pca, warning=FALSE, error=FALSE, message=FALSE}
experiment.aggregate <- RunPCA(object = experiment.aggregate, npcs=100)
```

Seurat then provides a number of ways to visualize the PCA results

Visualize PCA loadings
```{r viz_pca, warning=FALSE, error=FALSE, message=FALSE}
pca.loadings.1 <- VizDimLoadings(experiment.aggregate, dims = 1, ncol = 1) + theme_minimal(base_size = 8)
pca.loadings.2 <- VizDimLoadings(experiment.aggregate, dims = 2, ncol = 1) + theme_minimal(base_size = 8)

pca.loadings.1
pca.loadings.2

save_png_plot(plot = pca.loadings.1, filename = "pca_loadings_1.png")
save_png_plot(plot = pca.loadings.2, filename = "pca_loadings_2.png")
```

Principal components plot
```{r plot_pca, warning=FALSE, error=FALSE, message=FALSE}
pca.plot <- DimPlot(object = experiment.aggregate, reduction = "pca")

pca.plot

save_png_plot(plot = pca.plot, filename = "pca_dimplot.png")
```

Draws a heatmap focusing on a principal component. Both cells and genes are sorted by their principal component scores. Allows for nice visualization of sources of heterogeneity in the dataset.

```{r heatmap_pca, warning=FALSE, error=FALSE, message=FALSE}
pca.heatmap.1 <- DimHeatmap(object = experiment.aggregate, dims = 1:6, cells = 500, balanced = TRUE, fast = FALSE)

pca.heatmap.2 <- DimHeatmap(object = experiment.aggregate, dims = 7:12, cells = 500, balanced = TRUE, fast = FALSE)

pca.heatmap.1
pca.heatmap.2

save_png_plot(plot = pca.heatmap.1, filename = "pca_heatmap1.png")
save_png_plot(plot = pca.heatmap.2, filename = "pca_heatmap2.png")
```

### Selecting which PCs to use
To overcome the extensive technical noise in any single gene, Seurat clusters cells based on their PCA scores, with each PC essentially representing a metagene that combines information across a correlated gene set. Determining how many PCs to include downstream is therefore an important step.

ElbowPlot plots the standard deviations (or approximate singular values if running PCAFast) of the principle components for easy identification of an elbow in the graph. This elbow often corresponds well with the significant PCs and is much faster to run.  This is the traditional approach to selecting principal components.

```{r elbow, warning=FALSE, error=FALSE, message=FALSE}
elbow.plot <- ElbowPlot(experiment.aggregate, ndims = 100)

elbow.plot

save_png_plot(plot = elbow.plot, filename = "elbow_plot.png")
```
## Finally, lets save the filtered and normalized data
```{r save_rdata, warning=FALSE, error=FALSE, message=FALSE}
saveRDS(experiment.aggregate, file=part3.rds.save.path)

# to.save.for.report <- c("cell.cycle.table")

save(cell.cycle.table, file=part3.report.path)
```

## Session Information
```{r session_info, warning=FALSE, error=FALSE, message=FALSE}
sessionInfo()
```
